src/sage/matrix/matrix_mod2_dense.pyx

```diff
@@ -1919,6 +1919,74 @@ cdef class Matrix_mod2_dense(matrix_dense.Matrix_dense):   # dense or sparse
         self.cache('rank', r)
         return r
 
+    def _solve_right_general(self, B, check=True):
+        """
+        Solve the matrix equation AX = B for X using the M4RI library.
+
+        INPUT:
+
+        - ``B`` -- a matrix
+        - ``check`` -- boolean (default: ``True``); whether to check if the
+          matrix equation has a solution
+
+        EXAMPLES::
+
+            sage: A = matrix(GF(2), [[1, 0], [0, 1], [1, 1]])
+            sage: A.solve_right(vector([1, 1, 0]))
+            (1, 1)
+            sage: A.solve_right(vector([1, 1, 1]))
+            Traceback (most recent call last):
+            ...
+            ValueError: matrix equation has no solutions
+
+        TESTS::
+
+            sage: n = 128
+            sage: m = 128
+            sage: A = random_matrix(GF(2), n, m)
+            sage: B = A * random_vector(GF(2), m)
+            sage: A * A.solve_right(B) == B
+            True
+            sage: m = 64
+            sage: A = random_matrix(GF(2), n, m)
+            sage: B = A * random_vector(GF(2), m)
+            sage: A * A.solve_right(B) == B
+            True
+            sage: m = 256
+            sage: A = random_matrix(GF(2), n, m)
+            sage: B = A * random_vector(GF(2), m)
+            sage: A * A.solve_right(B) == B
+            True
+        """
+        cdef Matrix_mod2_dense X  # the solution
+
+        cdef mzd_t *B_entries = (<Matrix_mod2_dense>B)._entries
+        cdef rci_t rows = self._entries.nrows
+        if self._entries.nrows < self._entries.ncols:
+            rows = self._entries.ncols  # mzd_solve_left requires ncols <= nrows
+
+        cdef mzd_t *lhs = mzd_init(rows, self._entries.ncols)
+        mzd_copy(lhs, self._entries)
+        cdef mzd_t *rhs = mzd_init(rows, B_entries.ncols)
+        mzd_copy(rhs, B_entries)
+
+        sig_on()
+        # although it is called mzd_solve_left, it does the same thing as solve_right
+        ret = mzd_solve_left(lhs, rhs, 0, check)
+        sig_off()
+        mzd_free(lhs)
```

Comment:
Wouldn't this leak the `lhs` and `rhs` if user press ctrl+C?

Reply:
You are right, should I wrap it with a try...finally block or just disable the ability to Ctrl-C?

Reply:
Similarity, while `_right_kernel_matrix` does not enable signal but it can still leak the resulting kernel matrix when Ctrl-C (triggered during `self.new_matrix`).

Reply:
Certainly handling it properly is preferable. (But no memory leak is a hard thing to test for…)

cysignals is a bit complicated…

https://cysignals.readthedocs.io/en/latest/interrupt.html#using-sig-on-and-sig-off

So far I don't exactly understand what's the difference between `try: sig_on() except KeyboardInterrupt:` versus `if not sig_on_no_except():` either.

Reply:
for what it's worth, the part I don't understand is https://github.com/sagemath/cysignals/issues/205 .

In any case, I think the following will work:

```python
data = malloc(...)
if not sig_on_no_except():
	mzd_free(lhs)
	mzd_free(rhs)
	cython_check_exception()

ret = mzd_solve_left(...)
sig_off()
```

Reply:
Okay, I have pushed a version where try...finally are used to ensure allocated matrices get freed.